src/TorchSharp/Tensor/Storage.cs

```diff
@@ -0,0 +1,522 @@
+                if (i < 0 || i >= Count)
+                    throw new IndexOutOfRangeException();
+                return i;
+            }
+#endif
+
+            private int CheckIndex(int index) {
+                if (index < 0 || index >= Count)
+                    throw new IndexOutOfRangeException();
+                return index;
+            }
+
+            private void ValidateNotScalar()
+            {
+                if (_tensor is null &&_tensor_data_ptr == IntPtr.Zero)
+                    throw new InvalidOperationException("Invalid use of a scalar Storage instance.");
+            }
+
+            /// <summary>
+            /// The number of elements in the storage.
+            /// </summary>
+            public int Count {
+                get {
+                    ValidateNotScalar();
+                    if (_count == -1) {
+                        unsafe {
+                            _count = (int)nbytes() / sizeof(T);
+                        }
+                    }
+                    return _count;
+                }
+            }
+
+            private int _count = -1;
+
+            public static implicit operator Storage<T>(T value)
+            {
+                return new Storage<T>(value);
+            }
+
+            /// <summary>
+            /// Copy data from an array into a storage instance.
+            /// </summary>
+            /// <param name="array"></param>
+            public void copy_(IList<T> array)
+            {
+                ValidateNotScalar();
+                CopyFrom(array, 0);
+            }
+
+            /// <summary>
+            /// Fill a storage instance with a single value.
+            /// </summary>
+            /// <param name="value"></param>
+            public void fill_(T value)
+            {
+                ValidateNotScalar();
+                unsafe {
+                    var ptr = (T*)data_ptr();
+                    for (int i = 0; i < Count; i++)
+                        ((T*)ptr)[i] = value;
+                }
+            }
+
+            /// <summary>
+            /// Convert the storage to a .NET list.
+            /// </summary>
+            /// <returns></returns>
+            public IList<T> tolist()
+            {
+                ValidateNotScalar();
+                var result = new T[Count];
+                CopyTo(result, 0);
+                return result;
+            }
+
+            public bool IsReadOnly => false;
+
+            public bool Contains(T item)
+            {
+                ValidateNotScalar();
+                unsafe {
+                    var ptr = (T*)data_ptr();
+                    for (int i = 0; i < Count; i++)
+                        if (((T*)ptr)[i].Equals(item)) return true;
+                }
+                return false;
+            }
+
+            public void CopyTo(IList<T> array, int arrayIndex)
+            {
+                ValidateNotScalar();
+                int idx = arrayIndex;
```

Comment:
>idx [](http://example.com/codeflow?start=20&length=3)

Something looks wrong in this implementation. `idx` is defined and not used. also in the loop should `i` starts from 0?

I think this comment apply to CopyFrom too?